src/irs/adapters/inbound/s3.py

```diff
@@ -47,87 +48,104 @@ def calc_part_ranges(
     return part_ranges
 
 
-async def get_download_url(*, object_id: str, bucket_id: str) -> str:
-    """Get object download URL from S3 inbox bucket"""
-    storage = get_objectstorage()
-    return await storage.get_object_download_url(
-        bucket_id=bucket_id, object_id=object_id
-    )
+@dataclass
+class S3IDs:
+    """Container for bucket and object ID"""
 
+    bucket_id: str
+    object_id: str
 
-async def get_object_size(*, object_id: str, bucket_id: str) -> int:
-    """Get object size from S3 metadata"""
-    storage = get_objectstorage()
-    return await storage.get_object_size(bucket_id=bucket_id, object_id=object_id)
 
+class StagingHandlerPort(ABC):
+    """Abstract base for object storage functionality dealing with staging."""
 
-def get_objectstorage() -> S3ObjectStorage:
-    """Factoring this out makes it overridable by tests"""
-    config = S3Config(
-        s3_endpoint_url=CONFIG.s3_endpoint_url,
-        s3_access_key_id=CONFIG.s3_access_key_id,
-        s3_secret_access_key=CONFIG.s3_secret_access_key,
-    )
-    return S3ObjectStorage(config=config)
+    @abstractmethod
+    async def init_staging(self) -> None:
+        """Start staging a re-encrypted file to staging area, returns an upload id"""
 
+    @abstractmethod
+    async def stage_part(self, *, data: bytes, part_number: int) -> None:
+        """Save a file part to the staging area"""
 
-async def retrieve_parts(
-    *, url: str, object_size: int, part_size: int, offset: int = 0
-) -> AsyncGenerator[bytes, None]:
-    """Get all parts from inbox, starting with file content at offset"""
-    for start, stop in calc_part_ranges(
-        part_size=part_size, object_size=object_size, byte_offset=offset
-    ):
-        yield await retrieve_part(url=url, start=start, stop=stop)
+    @abstractmethod
+    async def complete_staging(self, *, parts: int) -> None:
+        """Complete the staging of a re-encrypted file"""
 
+    @abstractmethod
+    async def retrieve_parts(self, *, offset: int = 0) -> AsyncGenerator[bytes, None]:
+        """Get all parts from inbox, starting with file content at offset"""
 
-async def retrieve_part(*, url: str, start: int, stop: int) -> bytes:
-    """Get one part from inbox by range"""
-    try:
-        response = requests.get(
-            url=url, headers={"Range": f"bytes={start}-{stop}"}, timeout=60
+    @abstractmethod
+    async def retrieve_part(self, *, url: str, start: int, stop: int) -> bytes:
+        """Get one part from inbox by range"""
+
+
+class StagingHandler(StagingHandlerPort):
+    """Wrapper for object storage staging functionality."""
+
+    def __init__(
+        self,
+        object_storage: S3ObjectStorage,
+        inbox_ids: S3IDs,
```

Comment:
This information is coming with the events at runtime if so they should rather be part of the retrieve_part(s) method signatures.

Reply:
See comment below on StagingHandler creation